emscripten.py

```diff
@@ -228,7 +228,7 @@ def report_missing_symbols(all_implemented, pre):
   missing = list(set(shared.Settings.USER_EXPORTED_FUNCTIONS) - all_implemented)
 
   for requested in missing:
-    if ('function ' + asstr(requested)) in pre:
+    if ('function ' + asstr(requested) + '(') in pre:
```

Comment:
the code before this PR seems a little odd. is it looping over each unresolved symbols from wasm-ld, and looking if the string appears in the emitted JS? But the JS compiler emits errors and warnings for the symbols (and in a precise way) so the error has already been emitted, hasn't it?

When I create a small program with a missing symbol the error shows up and I never reach this line, so I'm missing some context I guess.

Separately, I'm confused why look in the full text. Can't we get this info from the JS compiler, something like this:
```diff
diff --git a/emscripten.py b/emscripten.py
index 294e7dfa1..9cf2aff96 100644
--- a/emscripten.py
+++ b/emscripten.py
@@ -215,28 +215,27 @@ class Memory():
 
 def apply_memory(js, memory):
   # Apply the statically-at-compile-time computed memory locations.
   # Write it all out
   js = js.replace('{{{ HEAP_BASE }}}', str(memory.dynamic_base))
   js = js.replace('{{{ STACK_BASE }}}', str(memory.stack_base))
   js = js.replace('{{{ STACK_MAX }}}', str(memory.stack_max))
   return js
 
 
-def report_missing_symbols(all_implemented, pre):
+def report_missing_symbols(all_implemented, forwarded_json):
   # the initial list of missing functions are that the user explicitly exported
   # but were not implemented in compiled code
   missing = list(set(shared.Settings.USER_EXPORTED_FUNCTIONS) - all_implemented)
 
   for requested in missing:
-    if ('function ' + asstr(requested)) in pre:
-      continue
+    if forwarded_json['Functions']['missingFunctions'].get(requested):
       diagnostics.warning('undefined', 'undefined exported symbol: "%s"', requested)
 
   # Special hanlding for the `_main` symbol
 
   if shared.Settings.STANDALONE_WASM:
     # standalone mode doesn't use main, and it always reports missing entry point at link time.
     # In this mode we never expect _main in the export list.
     return
 
   if shared.Settings.IGNORE_MISSING_MAIN:
@@ -369,21 +368,21 @@ def emscript(in_wasm, out_wasm, outfile_js, memfile, temp_files, DEBUG):
   # merge forwarded data
   shared.Settings.EXPORTED_FUNCTIONS = forwarded_json['EXPORTED_FUNCTIONS']
 
   exports = metadata['exports']
 
   if shared.Settings.ASYNCIFY:
     exports += ['asyncify_start_unwind', 'asyncify_stop_unwind', 'asyncify_start_rewind', 'asyncify_stop_rewind']
 
   all_exports = exports + list(metadata['namedGlobals'].keys())
   all_exports = set([asmjs_mangle(e) for e in all_exports])
-  report_missing_symbols(all_exports, pre)
+  report_missing_symbols(all_exports, forwarded_json)
 
   asm_consts = create_asm_consts(metadata)
   em_js_funcs = create_em_js(forwarded_json, metadata)
   asm_const_pairs = ['%s: %s' % (key, value) for key, value in asm_consts]
   asm_const_map = 'var ASM_CONSTS = {\n  ' + ',  \n '.join(asm_const_pairs) + '\n};\n'
   pre = pre.replace(
     '// === Body ===',
     ('// === Body ===\n\n' + asm_const_map +
      '\n'.join(em_js_funcs) + '\n'))
 
diff --git a/src/jsifier.js b/src/jsifier.js
index 8ca3d3533..71b5ea74b 100644
--- a/src/jsifier.js
+++ b/src/jsifier.js
@@ -133,20 +133,21 @@ function JSify(data, functionsOnly) {
           if (dependent) msg += ' (referenced by ' + dependent + ')';
           if (ERROR_ON_UNDEFINED_SYMBOLS) {
             error(msg);
             if (!LLD_REPORT_UNDEFINED) {
               warnOnce('Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols');
             }
             warnOnce('To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`')
             warnOnce(finalName + ' may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library')
           } else if (VERBOSE || WARN_ON_UNDEFINED_SYMBOLS) {
             warn(msg);
+            Functions.missingFunctions.push(ident);
           }
           if (ident === 'main' && STANDALONE_WASM) {
             warn('To build in STANDALONE_WASM mode without a main(), use emcc --no-entry');
           }
         }
         if (!RELOCATABLE) {
           // emit a stub that will fail at runtime
           LibraryManager.library[ident] = new Function("err('missing function: " + ident + "'); abort(-1);");
           // We have already warned/errored about this function, so for the purposes of Closure use, mute all type checks
           // regarding this function, marking ot a variadic function that can take in anything and return anything.
diff --git a/src/modules.js b/src/modules.js
index 82eb5c3bd..e6e8e55fb 100644
--- a/src/modules.js
+++ b/src/modules.js
@@ -37,20 +37,23 @@ function genArgSequence(n) {
     args.push('a'+i);
   }
   return args;
 }
 
 var Functions = {
   // All functions that will be implemented in this file. Maps id to signature
   implementedFunctions: {},
   // functions added from the library. value 2 means asmLibraryFunction
   libraryFunctions: {},
+  // Unresolved symbols, that is, functions that were expected to be found in
+  // the library but were missing.
+  missingFunctions: []
 };
 
 var LibraryManager = {
   library: null,
   structs: {},
   loaded: false,
   libraries: [],
 
   has: function(name) {
     return this.libraries.indexOf(name) >= 0;
```

(I can't test this because I'm not sure how to reach that code path.)